modP2pImpl/src/org/aion/p2p/impl1/tasks/TaskWrite.java

```diff
@@ -92,25 +92,50 @@ public void run() {
             }
             buf.flip();
 
+            long t1 = System.currentTimeMillis(), t2;
+            int wrote = 0;
             try {
-                while (buf.hasRemaining()) {
-                    // @Attention:  very important sleep , otherwise when NIO write buffer full,
-                    // without sleep will hangup this thread.
-                    Thread.sleep(0, 1);
-                    sc.write(buf);
+                do {
+                    int result = sc.write(buf);
+                    wrote += result;
+
+                    if (result == 0) {
+                        // @Attention:  very important sleep , otherwise when NIO write buffer full,
+                        // without sleep will hangup this thread.
+                        Thread.sleep(0, 1);
+                    }
+
+                    t2 = System.currentTimeMillis() - t1;
```

Comment:
It may be slightly safer to use System.nanoTime() for this measurement instead of System.currentTimeMillis().

curentTimeMillis changes if the system's clock changes, whereas nanoTime() does not: "This method can only be used to measure elapsed time and is not related to any other notion of system or wall-clock time."

https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#nanoTime--

Reply:
updated